src/julia_gc.c

```diff
@@ -589,6 +593,13 @@ static void PreGCHook(int full)
     // a thread-local variable.
     SaveTLS = JuliaTLS;
     JuliaTLS = jl_get_ptls_states();
+    // This is the same code as in VarsBeforeCollectBags() for GASMAN.
+    // It is necessary because ASS_LVAR() and related functionality
+    // does not call CHANGED_BAG() for performance reasons. CHANGED_BAG()
+    // is only called when the current lvars bag is being changed. Thus,
+    // we have to add a write barrier at the start of the GC, too.
+    if (STATE(CurrLVars))
+        CHANGED_BAG(STATE(CurrLVars));
```

Comment:
It's OK to have this in here for now, to get a quick fix; but on the long run, I prefer to fix this properly, by providing a generic "pre-gc hook" in both GASMAN (which already has one) and Julia GC, and then setting that to `VarsBeforeCollectBags`. Also, this "optimization" for lvars needs to be documented in a better place. But neither of these are change request for this PR; me or somebody else can take care of it in a future PR.

Reply:
I agree. My goal here was to get the fix out ASAP. As the code is identical to `VarsBeforeCollectBags()`, this should be unified.